tensorflow_model_optimization/python/core/sparsity_tf2/lthpruner_test.py

```diff
@@ -0,0 +1,486 @@
+        save_iteration=self.save_init,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+    init_weights_before_pruning = weight.read_value()
+
+    def _train():
+      p.create_slots(optimizer, weight)
+      p.preprocess_weights(optimizer, weight, self.grad(weight))
+      weight.assign(tf.math.add(weight, sample_noise(0))) # update weights
+      p.postprocess_weights(optimizer, weight, self.grad(weight))
+      optimizer.iterations.assign_add(1)
+      initialization_slot = optimizer.get_slot(weight, "original_initialization")
+      return initialization_slot
+    
+    initialization_slot = _train()
+    self.assertAllEqual(initialization_slot, init_weights_before_pruning)
+    
+    initialization_slot = tf.function(_train)()
+    self.assertAllEqual(initialization_slot, init_weights_before_pruning)
+
+
+  def testSaveWeightsIterK(self):
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+
+    save_round = 5
+    n_rounds = 24
+    end_iter = 100
+    frequency, prune_ratio_per_round = get_lth_sparsity(save_round, n_rounds, self.target_sparsity, end_iter)
+    pruning_schedule = make_pruning_schedule(1 - prune_ratio_per_round, save_round, end_iter, frequency)
+    
+    p = pruner.LTHPruner(
+        pruning_schedule=pruning_schedule,
+        save_iteration=save_round,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+
+    def _train(weight):
+      expected_saved_initialization = tf.ones_like(weight, dtype=weight.dtype.base_dtype) * -1
+      p.create_slots(optimizer, weight)
+      for i in range(7):
+        p.preprocess_weights(optimizer, weight, self.grad(weight))
+        if optimizer.iterations == save_round:
+          expected_saved_initialization = weight.read_value()
+        weight.assign(tf.math.add(weight, sample_noise(i)))
+        p.postprocess_weights(optimizer, weight, self.grad(weight))
+        optimizer.iterations.assign_add(1)
+      return expected_saved_initialization
+      
+    expected_saved_initialization = _train(weight)
+    self.assertAllEqual(optimizer.get_slot(weight, "original_initialization"), expected_saved_initialization)
+
+    initialization_slot_k = tf.math.multiply(optimizer.get_slot(weight, "original_initialization"), optimizer.get_slot(weight, "mask"))
+    masked_expected = tf.math.multiply(expected_saved_initialization, optimizer.get_slot(weight, "mask"))
+    self.assertAllEqual(initialization_slot_k, masked_expected)
+
+    mask_after_pruning = optimizer.get_slot(weight, "mask").read_value()
+    self.assertAllEqual(np.count_nonzero(mask_after_pruning), 97)
+
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+    optimizer.iterations.assign(0)
+    expected_saved_initialization = tf.function(_train)(weight)
```

Comment:
Let's make sure to create a brand-new optimizer (w/o any slot variables yet) instead of just setting the step count back to zero.

Make _train take the optimizer as an argument.